### PR TITLE
Have other agents wait for vault status(/persist/vault), before proceeding

### PR DESCRIPTION
**Pull request by cshari-zededa**
- So far, it is driven through device-steps.sh
- This PR is in a step towards pushing that logic inside agents
- downloader/verifier/volumemgr now wait for vault status before
  starting any work
- vaultmgr is still called in device-steps twice, that is to handle
  upgrade from old images where /persist/img and /persist/config
  may be encrypted i.e. setupDeprecatedVaults will be removed once
  all the devices move to a sentinel release (with this change)
- This PR also sets up the stage, where /persist/vault may only be
  unlocked after attestation is successful

Signed-off-by: Hariharasubramanian C S <cshari@zededa.com>

### CODE REVIEW

File: pkg/pillar/cmd/volumemgr/volumemgr.go [hunk 1]
```diff
@@ -111,11 +114,6 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
 	if err := pidfile.CheckAndCreatePidfile(log, agentName); err != nil {
 		log.Fatal(err)
 	}
-	log.Infof("Starting %s", agentName)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/volumemgr.go:114]: Why did you remove this line?

File: pkg/pillar/cmd/volumemgr/volumemgr.go [hunk 2]
```diff
@@ -430,6 +441,28 @@ func Run(ps *pubsub.PubSub, loggerArg *logrus.Logger, logArg *base.LogObject) in
...
+		select {
+		case change := <-subVaultStatus.MsgChan():
+			subVaultStatus.ProcessChange(change)
+		case <-stillRunning.C:
+		}
+		ps.StillRunning(agentName, warningTime, errorTime)
+	}
+	log.Infof("processed Vault Status")
+
+	log.Infof("Starting %s", agentName)
```
Comment (eriknordmark) [inline pkg/pillar/cmd/volumemgr/volumemgr.go:456]: Started a long time ago ...

File: pkg/pillar/scripts/device-steps.sh [hunk 3]
```diff
@@ -133,8 +133,8 @@ touch "$WATCHDOG_PID/zedbox.pid" "$WATCHDOG_FILE/zedbox.touch"
 
 if [ -c $TPM_DEVICE_PATH ] && ! [ -f $CONFIGDIR/disable-tpm ]; then
 #It is a device with TPM, enable disk encryption
-    if ! $BINDIR/vaultmgr setupVaults; then
-        echo "$(date -Ins -u) device-steps: vaultmgr setupVaults failed"
+    if ! $BINDIR/vaultmgr setupDeprecatedVaults; then
```
Comment (eriknordmark) [inline pkg/pillar/scripts/device-steps.sh:136]: Where do we setup up the non-deprecated vault?

**Review by eriknordmark — COMMENTED** [review]
Looking at this, the wait for GCInitialized, and the increasing addition of waiting for Onboarded, maybe we should take all of those and make them into a function in a package which does all of:
 - subscribe
 - handle the change
 - poke StillRunning
 - unsubscribe

Note that we need to add an unsubscribe to Pubsub for that, which would close the AF_UNIX socket etc
But a pattern with a waitForX() i.e., waitForVault() in this case, would mean less code duplication.

File: pkg/pillar/scripts/device-steps.sh (hunk 3 above)
Comment (cshari-zededa) [reply pkg/pillar/scripts/device-steps.sh:136]: I guess you are referring to /persist/vault handling. That is done when vaultmgr starts with "runAsService". The idea is that, at some point (when we no longer have /persist/config and /persist/img as vaults on any of the EVE devices), we can remove this block of code from device-steps.sh, and just retain the part where vaultmgr is started with runAsService (at that point, we no longer need that argument, we can simply start vaultmgr along with other agents).

**Comment by cshari-zededa** [comment]
> Looking at this, the wait for GCInitialized, and the increasing addition of waiting for Onboarded, maybe we should take all of those and make them into a function in a package which does all of:
> 
> * subscribe
> * handle the change
> * poke StillRunning
> * unsubscribe
> 
> Note that we need to add an unsubscribe to Pubsub for that, which would close the AF_UNIX socket etc
> But a pattern with a waitForX() i.e., waitForVault() in this case, would mean less code duplication.
> 
> That might make sense in the utils package; a utils/waitfor.go??

Sure, I can fold the current ones into utils/waitfor.go, and probably we can tackle unsubscribe as an incremental change in a follow-up PR?

File: pkg/pillar/cmd/volumemgr/volumemgr.go (hunk 2 above)
Comment (cshari-zededa) [reply pkg/pillar/cmd/volumemgr/volumemgr.go]: Corrected it

**Comment by cshari-zededa** [comment]
> Looking at this, the wait for GCInitialized, and the increasing addition of waiting for Onboarded, maybe we should take all of those and make them into a function in a package which does all of:
> 
> * subscribe
> * handle the change
> * poke StillRunning
> * unsubscribe
> 
> Note that we need to add an unsubscribe to Pubsub for that, which would close the AF_UNIX socket etc
> But a pattern with a waitForX() i.e., waitForVault() in this case, would mean less code duplication.
> 
> That might make sense in the utils package; a utils/waitfor.go??

I've introduced utils/waitfor.go, and updated the code in verifier/volumemgr/downloader to call this function.
Can you please take a look at the updated changes?

File: pkg/pillar/cmd/volumemgr/volumemgr.go (hunk 1 above)
Comment (cshari-zededa) [reply pkg/pillar/cmd/volumemgr/volumemgr.go:114]: Corrected it

File: pkg/pillar/utils/waitfor.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	return nil
+}
+
+func handleVaultStatusModify(ctxArg interface{}, key string,
+	statusArg interface{}) {
+
+	ctx := ctxArg.(*Context)
+	vault := statusArg.(types.VaultStatus)
+	if vault.Name == types.DefaultVaultName &&
+		vault.Status != info.DataSecAtRestStatus_DATASEC_AT_REST_ERROR {
```
Comment (eriknordmark) [inline pkg/pillar/utils/waitfor.go:62]: Interesting question what we might want to do if this fails. I guess zedagent will come up and report it. That's good enough for now.

File: pkg/pillar/cmd/vaultmgr/vaultmgr.go [hunk 5]
```diff
@@ -27,26 +28,26 @@ import (
 )
 
 type vaultMgrContext struct {
-	pubVaultStatus  pubsub.Publication
-	subGlobalConfig pubsub.Subscription
-	GCInitialized   bool // GlobalConfig initialized
+	pubVaultStatus       pubsub.Publication
+	subGlobalConfig      pubsub.Subscription
+	GCInitialized        bool // GlobalConfig initialized
+	defaultVaultUnlocked bool
```
[the remark this answers is not in the document]
Comment (cshari-zededa) [reply pkg/pillar/cmd/vaultmgr/vaultmgr.go:34]: This is just setting it up for the upcoming PR, where we will receive escrowed data back from zedagent in vaultmgr, and we will try unlocking the default vault, if it is not still unlocked.

[the remark this answers is not in the document]
Comment (eriknordmark) [reply pkg/pillar/cmd/vaultmgr/vaultmgr.go:34]: OK

**Review by eriknordmark — APPROVED** [review]
LGTM. Thanks for the utils function. I'll follow that pattern and add other utils.
